### ISSUE TITLE
Unable to make spawner alglomerations

### ISSUE THREAD

**Issue by dethbeginsagain** [CAD97/SpawnerCraft#20]
as i was working on my modpack, i noticed a few issues with the essence, when i tried to combine them in 1.12 they wouldn't combine and instead would try to duplicate themselves. Nothing popped up on the crafting grid, and i even installed the jei addon to test this and tried to craft it with its auto function, however as previous it failed. Any ideas to that? or is 1.12 not ready to be considered working just yet?

**Comment by CAD97** [comment]
1.12 is in a _very_ broken state; I just did the bare minimum to get it running, haven't tested it much (obviously.... 😛). That said, if you're using 1.12-4.0.0b1, I thought I did test it when I fixed #18.

I'm actually rewriting the mod properly (into [Kotlin](http://kotlinlang.org/)) to make it easier to work with, as the current code is old and messy with the changes Forge 1.12 brings. It still should work, however, so... this needs fixing. I'll be duplicating the recipe code over into the new codebase anyway so I need to figure out what went wrong.

```text
[21:06:33] [Server thread/FATAL]: Error executing task
java.util.concurrent.ExecutionException: java.lang.ArrayIndexOutOfBoundsException: 3
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:1.8.0_121]
	at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[?:1.8.0_121]
	at net.minecraft.util.Util.runTask(Util.java:54) [Util.class:?]
	at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:795) [MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:740) [MinecraftServer.class:?]
	at net.minecraft.server.integrated.IntegratedServer.tick(IntegratedServer.java:192) [IntegratedServer.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:589) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: java.lang.ArrayIndexOutOfBoundsException: 3
	at java.util.Arrays$ArrayList.get(Arrays.java:3841) ~[?:1.8.0_121]
	at net.minecraft.util.NonNullList.get(NonNullList.java:51) ~[NonNullList.class:?]
	at net.minecraft.item.crafting.ShapedRecipes.checkMatch(ShapedRecipes.java:121) ~[ShapedRecipes.class:?]
	at net.minecraft.item.crafting.ShapedRecipes.matches(ShapedRecipes.java:89) ~[ShapedRecipes.class:?]
	at cad97.spawnercraft.utility.NBTPreservingShapedRecipe.matches(NBTPreservingShapedRecipe.java:60) ~[NBTPreservingShapedRecipe.class:?]
	at net.minecraft.item.crafting.CraftingManager.findMatchingRecipe(CraftingManager.java:216) ~[CraftingManager.class:?]
	at net.minecraft.inventory.Container.slotChangedCraftingGrid(Container.java:883) ~[Container.class:?]
	at net.minecraft.inventory.ContainerWorkbench.onCraftMatrixChanged(ContainerWorkbench.java:54) ~[ContainerWorkbench.class:?]
	at net.minecraft.inventory.InventoryCrafting.setInventorySlotContents(InventoryCrafting.java:121) ~[InventoryCrafting.class:?]
	at net.minecraft.inventory.Slot.putStack(Slot.java:97) ~[Slot.class:?]
	at net.minecraft.inventory.Container.slotClick(Container.java:218) ~[Container.class:?]
	at net.minecraft.network.NetHandlerPlayServer.processClickWindow(NetHandlerPlayServer.java:1210) ~[NetHandlerPlayServer.class:?]
	at net.minecraft.network.play.client.CPacketClickWindow.processPacket(CPacketClickWindow.java:47) ~[CPacketClickWindow.class:?]
	at net.minecraft.network.play.client.CPacketClickWindow.processPacket(CPacketClickWindow.java:12) ~[CPacketClickWindow.class:?]
	at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:21) ~[PacketThreadUtil$1.class:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_121]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_121]
	at net.minecraft.util.Util.runTask(Util.java:53) ~[Util.class:?]
	... 5 more
```

**Comment by CAD97** [comment]
I think I need to submit some mappings to MCP, `net.minecraft.util.NonNullList::from` is _very_ misleading. Anyway, I'm uploading a new beta that fixes this.

I hope the build works now... 1.12 has really made a bigger mess of the code than it already was, hence the [rewrite in progress](https://github.com/CAD97/SpawnerCraft/tree/kt-1.12).

Though this testing does show me that I need to address the recipe book before releasing a non-beta build.
